### PR TITLE
Moving from Old Language Metadata Structure to Support Sub-languages and Simplified JSON

### PR DESCRIPTION
**Pull request by OmarAI2003**
### Contributor checklist

- [x] This pull request is on a [separate branch](https://docs.github.com/en/get-started/quickstart/github-flow) and not the main branch

---

### Description
This PR introduces several key updates and improvements to the project’s handling of language metadata, test cases, and language extraction logic:

1. **Refactored Language Metadata Structure**:
   - Simplified the `language_metadata.json` file by removing unnecessary nesting and keys, leaving only `language`, `iso`, and `qid` at the top level.
   - Organized sub-languages under their respective main languages (e.g., "Norwegian" now has "Nynorsk" and "Bokmål").

2. **New Language Support**:
   - Added new languages and dialects, including Mandarin, and others, into the `language_metadata.json` file.

3. **Improved Language Handling**:
   - Implemented a new `format_sublanguage_name` function to format sub-languages as `mainlang/sublang`.
   - Refactored the generation of `language_map` and `language_to_qid` to accommodate the new structure, ensuring correct directory creation and querying for sub-languages.
   - Removed the dependency on the obsolete `languages` key in the JSON file.

4. **Utility Functions and Minor Fixes**:
   - Refined `get_scribe_languages`, `list_all_languages`, and `_find` functions to handle both main languages and sub-languages.
   - Removed outdated utility functions (`get_language_words_to_remove`, `get_language_words_to_ignore`) no longer needed with the simplified JSON structure.
   - Enhanced error handling for cases where a language only has sub-languages.

5. **Updated Test Cases**:
   - Modified test cases to align with the new language metadata structure, including tests for sub-languages.
   - Added unit tests for sub-language formatting and listing, ensuring accurate results for all languages.

I’m open to feedback and welcome any additional suggestions or simple updates to improve the functionality and alignment with the project's overall goals.

### Related issue
This PR addresses **issue #293**

### CODE REVIEW

**Comment by andrewtavis** [comment]
CC @DeleMike, @catreedle, @Ekikereabasi-Nk, @KesharwaniArpita, @axif0 

Would be great to get your all's reviews!

**Comment by andrewtavis** [comment]
I'd like to have one check for queries and one check for structure, but aside from that I agree that we have a bit of a web going here and once we have the current PRs merged we'll have most of the functionality we're after :)

**Comment by DeleMike** [comment]
> I'd like to have one check for queries and one check for structure, but aside from that I agree that we have a bit of a web going here and once we have the current PRs merged we'll have most of the functionality we're after :)

alrighty!

**Comment by catreedle** [comment]
Will this help simplify issue #385, or does it already cover its functionality? If it does, could you please point out which part of the work addresses it? I'm still trying to understand how it connects, but I'm having difficulty. 😅

**Comment by OmarAI2003** [comment]
> Will this help simplify #385, or does it already cover its functionality? If it does, could you please point out which part of the work addresses it? I'm still trying to understand how it connects, but I'm having difficulty. 😅

"I think this should significantly simplify things, especially with the two newly added functions in `utils.py`. One function provides a list of all queryable languages from the JSON file, and the other formats language names appropriately. The formatted names are designed to be compatible with the `language_data_extraction` directory structure by returning main languages capitalized and sub-languages in the format `MainLang/SubLang`, with both parts capitalized."

File: src/scribe_data/resources/language_metadata.json [hunk 1]
```diff
@@ -1,70 +1,182 @@
...
+  },
+  "ukrainian": {
+    "iso": "ua",
+    "qid": "Q8798"
+  },
+  "yoruba": {
+    "iso": "yo",
+    "qid": "Q34311"
+  }
 }
```
Comment (KesharwaniArpita) [inline src/scribe_data/resources/language_metadata.json:182]: Thanks for cleaning the `language_metadata.py`. It is now much easier to look upto

**Comment by catreedle** [comment]
> > Will this help simplify #385, or does it already cover its functionality? If it does, could you please point out which part of the work addresses it? I'm still trying to understand how it connects, but I'm having difficulty. 😅
> 
> I think this should significantly simplify things, especially with the two newly added functions in `utils.py`. One function provides a list of all queryable languages from the JSON file, and the other formats language names appropriately. The formatted names are designed to be compatible with the `language_data_extraction` directory structure by returning main languages capitalized and sub-languages in the format `MainLang/SubLang`, with both parts capitalized.

Thank you! Will look into it. :)

File: tests/load/test_update_utils.py [hunk 2]
```diff
@@ -129,89 +164,69 @@ def test_get_language_from_iso_negative():
...
+        "tajik",
+        "tamil",
+        "ukrainian",
+        "yoruba",
+    ]
+
+    assert utils.list_all_languages() == expected_languages
 
 
 def test_get_ios_data_path():
```
Comment (KesharwaniArpita) [inline tests/load/test_update_utils.py:232]: The updated tests looks great! Thanks @OmarAI2003

File: src/scribe_data/cli/list.py [hunk 3]
```diff
@@ -83,8 +91,11 @@ def list_data_types(language: str = None) -> None:
...
         data_types = set()
-        for lang in language_metadata["languages"]:
-            language_dir = LANGUAGE_DATA_EXTRACTION_DIR / lang["language"].capitalize()
+        for lang in languages:
+            language_dir = (
+                LANGUAGE_DATA_EXTRACTION_DIR
+                / format_sublanguage_name(lang, language_metadata).capitalize()
+            )
             if language_dir.is_dir():
                 data_types.update(f.name for f in language_dir.iterdir() if f.is_dir())
```
Comment (KesharwaniArpita) [inline src/scribe_data/cli/list.py:101]: You've made sorting and formatting more efficient by calling list_all_languages() and sorting the languages using get_language_iso() and get_language_qid(). This reduces redundancy and enhances readability. Thank you for this!!!

**Review by KesharwaniArpita — COMMENTED** [review]
Thanks for your work @OmarAI2003 !!!!

File: src/scribe_data/resources/language_metadata.json (hunk 1 above)
Comment (OmarAI2003) [reply src/scribe_data/resources/language_metadata.json:182]: Thank you for your feedback! I'm glad to hear that the JSON file is easier to work with now. I also want to thank @andrewtavis for his guidance on this.

**Review by andrewtavis — APPROVED** [review]
Really really great work here, @OmarAI2003 😊 This is going to make so many parts of Scribe-Data so much easier :) Appreciate your efforts here!

**Comment by OmarAI2003** [comment]
Thank you! @andrewtavis  😊 I really appreciate your guidance throughout this process. I'm excited to contribute even more and keep improving Scribe!

**Pull request merged**